### ISSUE TITLE
Self attention computation

### ISSUE THREAD

**Issue by lxuechen** [lucidrains/denoising-diffusion-pytorch#3]
Hi, thanks for porting the tf project to torch and open sourcing the code. 

Looking at the linear attention code, it seems quite different from the standard self attention in at least two ways:
- the position of [softmax](https://github.com/lucidrains/denoising-diffusion-pytorch/blob/5989f4c77eafcdc6be0fb4739f0f277a6dd7f7d8/denoising_diffusion_pytorch/denoising_diffusion_pytorch.py#L175) is different from the usual softmax after computing QK^T. 
- [summation](https://github.com/lucidrains/denoising-diffusion-pytorch/blob/5989f4c77eafcdc6be0fb4739f0f277a6dd7f7d8/denoising_diffusion_pytorch/denoising_diffusion_pytorch.py#L176) is over the w*h dimension, as opposed to the usual channel dimension

**Comment by lxuechen** [comment]
Ok I realized that this implementation is based on linear complexity self attention.

**Comment by lucidrains** [comment]
Indeed

**Comment by lxuechen** [comment]
This resolves the softmax position issue, but it doesn't fully explain why the summation is over the (w * h) dimension, as opposed to the channel dimension. Summing over the channels might make more sense, since the attention matrix would then be of size (w * h) * (w * h), which correspondes to attending over the features spatially.

**Comment by lucidrains** [comment]
@lxuechen I think this paper will have more details about this particular flavor of attention https://arxiv.org/abs/1812.01243

**Comment by lucidrains** [comment]
@lxuechen basically, in short, it is an order of operations change, from (qk)v -> q(kv)

**Comment by lucidrains** [comment]
it is also the basis for new work to try to undercut the quadratic cost in NLP https://github.com/lucidrains/performer-pytorch and https://github.com/lucidrains/linear-attention-transformer

**Comment by lxuechen** [comment]
> basically, in short, it is an order of operations change, from (qk)v -> q(kv)

Ok, this makes it clear. It saves compute when (h * w) < c.
